foca/security/auth.py

```diff
@@ -0,0 +1,613 @@
+"""Decorator and utility functions for protecting access to endpoints."""
+
+from connexion.exceptions import Unauthorized
+from connexion import request
+from functools import wraps
+import logging
+from typing import (Callable, Iterable, Mapping, Optional)
+
+from cryptography.hazmat.primitives import serialization
+import jwt
+import requests
+import json
+
+from foca.config.config_parser import get_conf, get_conf_type
+
+
+# Get logger instance
+logger = logging.getLogger(__name__)
+
+
+def param_pass(config_dict):
+    def auth_token_optional(fn: Callable) -> Callable:
+        """
+        **The decorator protects an endpoint from being called without a valid
+        authorization token.
+        """
+        @wraps(fn)
+        def wrapper(*args, **kwargs):
+
+            # Check if authentication is enabled
+            if get_conf(
+                config_dict,
+                'security',
+                'authorization_required',
+            ):
+
+                # Get config parameters
+                validation_methods = get_conf_type(
```

Comment:
All of these should be passed through to the decorator from the calling context. Otherwise all apps using FOCA must define their configs _exactly_ int the way we do here, with the same categories and subcategories. This is not really nice. Better to do all those `get_conf()` calls in the calling context and passing them to the decorator, because it is much more generic. However, to make sure the decorator is still simple to use, we should set reasonable defaults, so that only the minimum number of parameters needs to be passed.

Reply:
So what you are suggesting is to pass the corresponding values of those given functions as a parameter via decorator instead of passing `config` as an object instance?

Reply:
What I am thinking is to create a directory of the respective function return values and the pass it as a parameter

Reply:
From an API design perspective, I think I would prefer the option where each parameter is passed individually, by name, because it's cleaner (named args in Python functions are specifically designed for that use case) than passing an opaque dictionary (how/where do you define that dictionary? how do you verify it, e.g., in terms of typing?). From the calling context perspective, if one desires, one could still pass a dictionary with `**d`, which would then get expanded to each named parameter (perhaps it would be good to rename `expected_prefix` in the decorator code with `token_prefix` so that we wouldn't have to rename any params if passing a dict from the config object). I'm not totally opposed to the dictionary option though as it does the trick as well.

But yes, values should be passed to the decorator, and missing values should take default values.  :-)

Reply:
I commented that before we discussed it over slack :p